metrics/density/docker_memory_usage.sh

```diff
@@ -0,0 +1,318 @@
+		if (( newscan != oldscan )); then
+			echo -e "\nnew full_scan ($oldscan to $newscan)"
+
+			newpages=$(cat /sys/kernel/mm/ksm/pages_shared)
+			# Do we have a previous scan to compare with
+			echo "check pages $oldpages to $newpages"
+			if (( oldpages != -1 )); then
+				# avoid divide by zero problems
+				if (( $oldpages > 0 )); then
+					pcnt=$(( 100 - ((newpages * 100) / oldpages) ))
+					# abs()
+					pcnt=$(( $pcnt * -1 ))
+
+					echo "$oldpages to $newpages is ${pcnt}%"
+
+					if (( $pcnt <= 5 )); then
+						echo "KSM stabilised at ${t}s"
+						return
+					fi
+				else
+					echo "$oldpages KSM pages... waiting"
+				fi
+			fi
+			oldscan=$newscan
+			oldpages=$newpages
+		else
+			echo -n "."
+		fi
+		sleep 1
+	done
+	echo "Timed out after ${1}s waiting for KSM to settle"
+}
+
+# Try to work out the 'average memory footprint' of a container.
+get_docker_memory_usage(){
+	hypervisor_mem=0
+	shim_mem=0
+	proxy_mem=0
+	proxy_mem=0
+	memory_usage=0
+
+	containers=()
+
+	for ((i=1; i<= NUM_CONTAINERS; i++)); do
+		containers+=($(random_name))
+		${DOCKER_EXE} run --runtime "$RUNTIME" --name ${containers[-1]} -tid $IMAGE $CMD
+	done
+
+	if [ "$AUTO_MODE" == "auto" ]; then
+		if (( ksm_on != 1 )); then
+			die "KSM not enabled, cannot use auto mode"
+		fi
+
+		echo "Entering KSM settle auto detect mode..."
+		wait_ksm_settle $WAIT_TIME
+	else
+		# If KSM is enabled, then you normally want to sleep long enough to
+		# let it do its work and for the numbers to 'settle'.
+		echo "napping $WAIT_TIME s"
+		sleep "$WAIT_TIME"
+	fi
+
+	metrics_json_start_array
+	# Check the runtime in order in order to determine which process will
+	# be measured about PSS
+	if [ "$RUNTIME" == "runc" ]; then
+		runc_workload_mem="$(get_runc_pss_memory)"
+		memory_usage="$runc_workload_mem"
+
+	local json="$(cat << EOF
+	{
+		"average": {
+			"Result": $memory_usage,
+			"Units" : "KB"
+		},
+		"runc": {
+			"Result": $runc_workload_mem,
+			"Units" : "KB"
+		}
+	}
+EOF
+)"
+
+	elif [ "$RUNTIME" == "cor" ] || [ "$RUNTIME" == "cc-runtime" ] || [ "$RUNTIME" == "kata-runtime" ]; then
+		# Get PSS memory of CC components.
+		# And check that the smem search has found the process - we get a "0"
+		#  back if that procedure fails (such as if a process has changed its name
+		#  or is not running when expected to be so)
+		# As an added bonus - this script must be run as root (or at least as
+		#  a user with enough rights to allow smem to read the smap stats for
+		#  the docker owned processes). Now if you do not have enough rights
+		#  the smem failure to read the stats will also be trapped.
+
+		hypervisor_mem="$(get_pss_memory "$HYPERVISOR_PATH")"
+		if [ "$hypervisor_mem" == "0" ]; then
+			die "Failed to find PSS for $HYPERVISOR_PATH"
+		fi
+
+		shim_mem="$(get_pss_memory "$CC_SHIM_PATH")"
+		if [ "$shim_mem" == "0" ]; then
+			die "Failed to find PSS for $CC_SHIM_PATH"
+		fi
+
+		proxy_mem="$(get_pss_memory "$CC_PROXY_PATH")"
```

Comment:
`PROXY_PATH?`

Reply:
Good spots on the CC stuff - I'll have a look to see why I didn't spot that failing during running...

Reply:
and... the reason it did not fail seems to be a bit of a quirk of `smem`, that didn't fail if there was no argument to its `-P`. I've added an extra check now so if we fail to expand a path for the vm/shim/proxy etc. then we fail and abort the test (at the point where we try to measure their PSS).